---
title: "assignment"
author: "Guillermo Turiel"
date: "2023-05-22"
output: html_document
---

```{r}
suppressWarnings(suppressPackageStartupMessages({
  library(GenomicRanges)
  library(epiwraps)
  library(ggplot2)
  library(rGREAT)
  library(AnnotationHub)
  library(ensembldb)
  library(bsseq)
  library(BiocParallel)
  library(edgeR)
  library(DMRcate)
  library(rtracklayer)
  library(sechm)
  library(pheatmap)
  library(viridis)
  library(data.table)
}))
proxy_url <- "http://proxy.ethz.ch:3128";
Sys.setenv(http_proxy = proxy_url, https_proxy = proxy_url, ftp_proxy = proxy_url)

setwd("C:/Users/guruiz/Desktop/ethEpigen2023/week11")

set.seed(40)
```

## Load the data

Download:
```{r, download}
options(timeout = 6000)
download.file("https://ethz-ins.org/content/w11_practical.zip", "w11_practical.zip")
unzip("w11_practical.zip")
bs <- readRDS("bs.rds")
```

## Testing

Get annotations (hs):
```{r}
# genes
ah <- AnnotationHub()
ensdb <- ah[["AH109336"]]

chr22 <-  GRanges(seqnames=Rle(c("22")), 
                  ranges = IRanges(1, end=195471971))
genesChr22 <- genes(ensdb, columns=c("gene_seq_start", "gene_seq_end", "gene_name"),
                    filter=GRangesFilter(chr22))

```


Differential methylation testing: 
```{r}
seqlevelsStyle(bs) <- "UCSC"
design <- model.matrix(~Type+Pair, data=pData(bs)) # design matrix

# adapt for methylation data
methdesign <- modelMatrixMeth(design)

seqAnnot <- sequencing.annotate(bs, methdesign, all.cov=TRUE, coef="Typenormal")
dmrcateRes <- dmrcate(seqAnnot, C=2, min.cpgs = 10, pcutoff=0.01)

dmrRanges <- extractRanges(dmrcateRes, genome="hg38")

dmrRangesGenes <- dmrRanges[!is.na(dmrRanges$overlapping.genes)]
```
Obtain the coordinates of the genes within DMRs. 
```{r}
# Get the genes within top 5 Differentially methylated regions
topIdx <- order(dmrRangesGenes$min_smoothed_fdr)[1:4]
genesDmr <- unlist(tstrsplit(dmrRangesGenes[topIdx]$overlapping.genes, split=", "))
genesDmr <- genesDmr[!is.na(genesDmr)]
dmrGenes <- genesChr22[genesChr22$gene_name %in% genesDmr]
dmrGenes
```
Retrieve methylation levels and visualize:
```{r}
seqlevelsStyle(bs) <- "NCBI"
metPr <- bsseq::getMeth(bs, 
                        regions=genesChr22, 
                        what="perRegion")
colnames(metPr) <- colnames(bs)
rownames(metPr) <- genesChr22$gene_name
metPr <- metPr[!is.na(rowSums(metPr)),]

library(viridis)
library(pheatmap)
annotationCol <- as.data.frame(pData(bs)[,c("Type","Pair")])
rownames(annotationCol) <- colnames(metPr)
pheatmap::pheatmap(metPr[dmrGenes$gene_name,], 
                   cluster_rows=TRUE,
                   cluster_cols=FALSE,
                   annotation_col=annotationCol,
                   show_rownames = TRUE,
                   color=rocket(10))
```

Extracting significant DMRs in normal and cancer samples:

```{r}
normal_dmr <- dmrRangesGenes[dmrRangesGenes$min_smoothed_fdr < 0.05 & dmrRangesGenes$meandiff > 0,]
seqlevelsStyle(normal_dmr) <- "UCSC"
cancer_dmr <- dmrRangesGenes[dmrRangesGenes$min_smoothed_fdr < 0.05 & dmrRangesGenes$meandiff < 0,]
seqlevelsStyle(cancer_dmr) <- "UCSC"

#seqlevelsStyle(genesChr22) <- "UCSC"
#Previous line gives this error
#Error in find_NCBI_assembly_ftp_dir(assembly_accession, assembly_name = assembly_name) :
#unable to find FTP dir for assembly GCF_000001405.40 in https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/
```

I couldnt solve the previous error so I am unable to continue. Anyway the code should follow like this:

```{r}
#job1 <- submitGreatJob(gr=normal_dmr, bg=genesChr22, species="hg38")
#res1 <- getEnrichmentTables(job1)

#job2 <- submitGreatJob(gr=cancer_dmr, bg=genesChr22, species="hg38")
#res2 <- getEnrichmentTables(job2)
```


We plot the top Molecular Functions:

```{r, fig.width=9, fig.height=6}
#bp1 <- res1$`GO Molecular Function`
#bp2 <- res2$`GO Molecular Function`


#ggplot(head(bp1,15), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
#ggplot(head(bp2,15), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c()
```


Like this, we are testing whether significant DMRs (FDR < 0.05) in normal (normal_dmr object) or cancer (cancer_dmr object) samples are enriched for some specific processes compared to the rest of genes of chr22. Then we will extract the "molecular function" processes from the enrichment analysis and plot the top ones using ggplot(). This we way we could see which processes are enriched in normal or cancer samples. 
